data/txDupCache.go

```diff
@@ -0,0 +1,256 @@
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package data
+
+import (
+	"context"
+	"encoding/binary"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-deadlock"
+
+	"golang.org/x/crypto/blake2b"
+)
+
+// digestCache is a rotating cache of size N accepting crypto.Digest as a key
+// and keeping up to 2*N elements in memory
+type digestCache struct {
+	cur  map[crypto.Digest]struct{}
+	prev map[crypto.Digest]struct{}
+
+	maxSize int
+	mu      deadlock.RWMutex
+}
+
+func makeDigestCache(size int) *digestCache {
+	c := &digestCache{
+		cur:     map[crypto.Digest]struct{}{},
+		maxSize: size,
+	}
+	return c
+}
+
+// check if digest d is in a cache.
+// locking semantic: write lock must be taken
+func (c *digestCache) check(d *crypto.Digest) bool {
+	_, found := c.cur[*d]
+	if !found {
+		_, found = c.prev[*d]
+	}
+	return found
+}
+
+// swap rotates cache pages.
+// locking semantic: write lock must be taken
+func (c *digestCache) swap() {
+	c.prev = c.cur
+	c.cur = map[crypto.Digest]struct{}{}
+}
+
+// put adds digest d into a cache.
+// locking semantic: write lock must be taken
+func (c *digestCache) put(d *crypto.Digest) {
+	if len(c.cur) >= c.maxSize {
+		c.swap()
+	}
+	c.cur[*d] = struct{}{}
+}
+
+// CheckAndPut adds digest d into a cache if not found
+func (c *digestCache) CheckAndPut(d *crypto.Digest) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.check(d) {
+		return true
+	}
+	c.put(d)
+	return false
+}
+
+// Len returns size of a cache
+func (c *digestCache) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return len(c.cur) + len(c.prev)
+}
+
+// Delete from the cache
+func (c *digestCache) Delete(d *crypto.Digest) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.cur, *d)
+	delete(c.prev, *d)
+}
+
+// txSaltedCache is a digest cache with a rotating salt
+// uses blake2b hash function
+type txSaltedCache struct {
+	digestCache
+
+	curSalt  [4]byte
+	prevSalt [4]byte
+	ctx      context.Context
+}
+
+func makeSaltedCache(size int) *txSaltedCache {
+	return &txSaltedCache{
+		digestCache: *makeDigestCache(size),
+	}
+}
+
+func (c *txSaltedCache) start(ctx context.Context, refreshInterval time.Duration) {
+	c.ctx = ctx
+	if refreshInterval != 0 {
+		go c.salter(refreshInterval)
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.moreSalt()
+}
+
+// salter is a goroutine refreshing the cache by schedule
+func (c *txSaltedCache) salter(refreshInterval time.Duration) {
+	ticker := time.NewTicker(refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			c.Remix()
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+// moreSalt updates salt value used for hashing
+func (c *txSaltedCache) moreSalt() {
+	r := uint32(crypto.RandUint64() % math.MaxUint32)
+	binary.LittleEndian.PutUint32(c.curSalt[:], r)
+}
+
+// Remix is a locked version of innerSwap, called on schedule
+func (c *txSaltedCache) Remix() {
```

Comment:
Do we need `Remix` exported out of the package?

Reply:
It is your call for what gets exported what not. 
On the one hand, everything here is only used in the package, so no need to export anything. 
On the other hand, some are internal interfaces of the cache others are the front facing ones which are also exported.

Reply:
I made CheckAndPut, Remix, Delete exported to define the interface more-less clearly